pyigm/clustering/survey_2d2pcf.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import print_function, absolute_import, division, unicode_literals
+
+import numpy as np
+from scipy.ndimage import gaussian_filter as gf
+
+from pyigm.clustering.xcorr_utils import random_gal, auto_pairs_rt, cross_pairs_rt, W3, collapse_along_LOS
+
+class Survey2D2PCF(object):
+    """
+    Class for 2D-2PCF for absorber-galaxy two-point
+    cross and auto-correlations. The Class loads several independent
+    IgmGalaxyField objects together. It internally calculates the total numbers
+    of cross- and auto-pairs of real data and produce randoms. It calculates
+    the 2D2PCF using the Landy & Szalay estimator (or others), and it estimates
+    the uncertainty with a bootstrap, jacknife or Landy & Szalay approximation.
+
+    Parameters
+    ----------
+    field : ClusteringField object
+      The field which initializes sets the binning.  It also
+      sets whether galaxy-galaxy, absorber-absorber, and galaxy-absorber analysis is on
+      So, load a good one first!
+    """
+
+    def __init__(self, field):
+        #f = copy.deepcopy(field)
+        self.fields = [field]
+
+        if field.galreal is not None:
+            self.DgDg = field.DgDg
+            self.DgRg = field.DgRg
+            self.RgRg = field.RgRg
+            self.Ngal = len(field.galreal)
+            self.gal_anly = True
+        else:
+            self.gal_anly = False
+
+        if field.absreal is not None:
+            self.DaDa = field.DaDa
+            self.DaRa = field.DaRa
+            self.RaRa = field.RaRa
+            self.Nabs = len(field.absreal)
+            self.abs_anly = True
+            # Both!
+            if field.galreal is not None:
+                self.DaDg = field.DaDg
+                self.DaRg = field.DaRg
+                self.RaDg = field.RaDg
+                self.RaRg = field.RaRg
+        else:
+            self.abs_anly = False
+
+
+        # Field edges
+        self.rbinedges = field.rbinedges # Should check that any new field has the same edges
+        self.tbinedges = field.tbinedges
+
+        # Binning
+        self.rcbins = 0.5*(self.rbinedges[:-1] + self.rbinedges[1:])
+        self.tcbins = 0.5*(self.tbinedges[:-1] + self.tbinedges[1:])
+        self.rdiff  = self.rbinedges[1:] - self.rbinedges[:-1]
+        self.tdiff  = self.tbinedges[1:] - self.tbinedges[:-1]
+
+    def addField(self, new_field):
+        """ Add a new field
+
+        Parameters
+        ----------
+        new_field : ClusterField
+
+        Returns
+        -------
+        All internal appends
+
+        """
+        # Check edges
+        if not np.isclose(new_field.rbinedges, self.rbinedges):
+            raise IOError("rbinedges of new field does not match Survey!!")
+        if not np.isclose(new_field.tbinedges, self.tbinedges):
+            raise IOError("tbinedges of new field does not match Survey!!")
+
+        #f = copy.deepcopy(field)
+        self.fields.append(new_field)
+
+        if new_field.galreal is not None:
```

Comment:
The fields must be independent for this to make sense. Should implement a check on coordinate to ensure the user is not messing up.

Reply:
Have added a test with 10" tolerance.

Reply:
10" seems small, I would go for 0.5 deg. If the user know what they're doing they can reduce the tolerance.

Reply:
ok, extending to 0.5 deg and is now
a user-supplied value.

Reply:
I should have said, optionally supplied